From: Fedora/Cygwin
Subject: Fix echo_input_at_read in command substitution

When set -v is active and a command substitution is executed,
the substituted command would be echoed multiple times. This patch
saves and restores echo_input_at_read around command substitutions
to prevent duplicate echoing.

--- a/parse.y
+++ b/parse.y
@@ -4509,6 +4509,8 @@ xparse_dolparen (
   sflags |= SEVAL_NOLONGJMP;
   save_parser_state (&ps);
   save_input_line_state (&ls);
+  /* avoid echoing every substitution again */
+  echo_input_at_read = 0;
   orig_eof_token = shell_eof_token;
 #if defined (ALIAS) || defined (DPAREN_ARITHMETIC)
   saved_pushed_strings = pushed_string_list;	/* separate parsing context */
--- a/subst.c
+++ b/subst.c
@@ -9707,6 +9707,7 @@ param_expand (
   WORD_LIST *list, *l;
   WORD_DESC *tdesc, *ret;
   int tflag, nullarg;
+  int old_echo_input;
 
 /*itrace("param_expand: `%s' pflags = %d", string+*sindex, pflags);*/
   zindex = *sindex;
@@ -10105,6 +10106,9 @@ arithsub:
 	}
 
 comsub:
+      old_echo_input = echo_input_at_read;
+      /* avoid echoing every substitution again */
+      echo_input_at_read = 0;
       if (pflags & PF_NOCOMSUB)
 	/* we need zindex+1 because string[zindex] == RPAREN */
 	temp1 = substring (string, *sindex, zindex+1);
@@ -10117,6 +10121,7 @@ comsub:
 	}
       FREE (temp);
       temp = temp1;
+      echo_input_at_read = old_echo_input;
       break;
 
     /* Do POSIX.2d9-style arithmetic substitution.  This will probably go
